### ISSUE TITLE
Clarification on interface grammar with struct

### ISSUE THREAD

**Issue by chergert** [varlink/varlink.github.io#31]
Hey!

I've been playing around writing an implementation of this to get the feel for things, and had a question while writing the interface parser.

We have:

```
"method" name struct -> struct
```

And I just wanted clarification that you cannot have:

```
"method" name type -> type
```

e.g. recursive decent parsing of method need-not carry a type-table of types defined before it.

**Comment by allisonkarlitskaya** [comment]
I feel like this would result in something kinda weird happening: if we support any type we'd also probably be considering "string", "bool", etc. in addition to the usual "struct"-flavoured custom types.  That means we'd have a field called `"parameters":` on the wire which could then be a boolean or a string or something else, which seems a bit weird.

Indeed, this field is documented

```
  parameters: ?object,
```

which is a pretty clear indication that it must be an object.  At least for input.

You could restrict to "object-like" types, which would mean the introduction of a new grammar category for something that might be a "vstruct" or might be a typedef name for a struct-like (ie: not enum) type, which is what I think you're asking about.

I agree that this is a more interesting question for return values... most programming languages have rather nice support for named input arguments for functions but multiple return values end up feeling a bit strange...

But it seems clear to me from reading the spec that the short answer to your question is definitely "no.".  It's very clear that it must be an inline struct, contained within `(` `)`.

**Comment by emersion** [comment]
Yeah, my understanding is that only structs are allowed. (This is important for extensibility, among other things.)
